### PR TITLE
move targetOS from global to target

### PR DESCRIPTION
**Pull request by thewilsonator**
part of a broader effort to move target specific and target dependant parameters from `global.params` to `target` to facilitate cross compilation

### CODE REVIEW

File: src/dmd/target.h [hunk 1]
```diff
@@ -74,6 +74,26 @@ struct TargetObjC
...
+        /* These are mutually exclusive; one and only one is set.
+         * Match spelling and casing of corresponding version identifiers
+         */
+        TargetOS_linux        = 1,
+        TargetOS_Windows      = 2,
+        TargetOS_OSX          = 4,
+        TargetOS_OpenBSD      = 8,
+        TargetOS_FreeBSD      = 0x10,
+        TargetOS_Solaris      = 0x20,
+        TargetOS_DragonFlyBSD = 0x40,
```
Comment (ibuclaw) [inline src/dmd/target.h:89]: Don't need the Target prefix here, just OS_xxx

Comment (thewilsonator) [reply src/dmd/target.h]: done

File: src/dmd/astbase.d [hunk 2]
```diff
@@ -6810,13 +6810,27 @@ struct ASTBase
...
+            OSX          = 4,
+            OpenBSD      = 8,
+            FreeBSD      = 0x10,
+            Solaris      = 0x20,
+            DragonFlyBSD = 0x40,
+            all = linux | Windows | OSX | OpenBSD | FreeBSD | Solaris | DragonFlyBSD,
+            Posix = linux | OSX | OpenBSD | FreeBSD | Solaris | DragonFlyBSD,
+        }
+
+        extern (C++) __gshared OS os;
```
Comment (thewilsonator) [inline src/dmd/astbase.d:6826]: @RazvanN7 is this the correct way to do things here?

File: src/dmd/e2ir.d [hunk 3]
```diff
@@ -1945,7 +1945,7 @@ elem *toElem(Expression e, IRState *irs)
                     elem *earray = ExpressionsToStaticArray(ne.loc, ne.arguments, &sdata);
 
                     e = el_pair(TYdarray, el_long(TYsize_t, ne.arguments.dim), el_ptr(sdata));
-                    if (irs.params.targetOS == TargetOS.Windows && irs.params.is64bit)
+                    if (target.os == Target.OS.Windows && irs.params.is64bit)
```
Comment (ibuclaw) [inline src/dmd/e2ir.d:1948]: More broadly speaking, these conditions should be replaced by a target method/hook for better cross compilation support.

For instance this cryptic condition and other similar ones for calling addressElem seem to be for determining whether the target passes dynamic arrays by reference. So they could all potentially be replaced with `target.passDArrayByRef` or `target.passByReference(ty)`.

Once all have been converted, OS and CPU information can just be made private.

File: src/dmd/astbase.d [hunk 4]
```diff
@@ -6810,13 +6810,27 @@ struct ASTBase
...
         extern (C++) static Type va_listType(const ref Loc loc, Scope* sc)
```
Comment (RazvanN7) [inline src/dmd/astbase.d:6827]: I think that this function should be here: the parser does not use it and it isn't used internally by ASTBase. You could try deleting it, that way you do not have to duplicate the code.

**Review by RazvanN7 — APPROVED** [review]

**Comment by WalterBright** [comment]
Can we please stop quickly merging large refactorings in just a few hours? Do the 72hr thing instead? There is no urgency to these refactorings, they don't fix anything.

Much of the glue code was set up to avoid global references by using the `irs` parameter. This PR regresses that making things worse.

**Comment by thewilsonator** [comment]
> Can we please stop quickly merging large refactorings in just a few hours? Do the 72hr thing instead? 

Please  no, also 10 is not a few hours. Perhaps this is more nuanced that we are 8h apart but there were approving reviews. This is a move of a singular symbol this is as small as is can possibly be.  So is https://github.com/dlang/dmd/pull/12367 which is 3 because of a import/(cyclical) module constructor dependancy which I discovered while trying to make it one symbol. You can have small PR (which this is) and fast, or you can have large PR or no contributions at all. If you don't like this I can go back to PR that get all this stuff done at once...

> There is no urgency to these refactorings, they don't fix anything.

... which does fix things. I do not have the time to go through 72h a pop for a series of refactoring PR. I can wait 72 for a monolithic but not for each in a series.

> Much of the glue code was set up to avoid global references by using the irs parameter. 

https://github.com/dlang/dmd/pull/12391 do you want to wait 72h for that? I don't I have more PRs that will have merge conflicts with that. I'm happy to fix things up and revert bad PRs but I'm not going to wait 3 days for a one symbol refactor.

**Comment by ibuclaw** [comment]
I'll just repeat an earlier comment that got hidden away by being "resolved".

---

More broadly speaking, these conditions should be replaced by a target method/hook for better cross compilation support.

For instance cryptic conditions like... 
```
if (irs.params.os == TargetOS.Windows && irs.params.is64bit)
```
... and other similar ones for calling `addressElem` seem to be for determining whether the target passes dynamic arrays by reference. So they could all potentially be replaced with `target.passDArrayByRef` or `target.passByReference(ty)`.

Once all have been converted, OS and CPU information can just be made private.

---

The point of the comment is that although it would take many more PRs to get done, this change would be only dozens of lines to move the fields from params to target if all users were fixed first to not depend on them in the first place.

**Comment by ibuclaw** [comment]
This PR caused a bootstrap regression https://issues.dlang.org/show_bug.cgi?id=21813

**Pull request merged**